numpy/lib/index_tricks.py

```diff
@@ -596,6 +597,419 @@ def __iter__(self):
     next = __next__
 
 
+class ndrange(_nx.collections_abc.Sequence):
+    """
+    The ndrange type represents an immutable sequence of numbers and can be
+    used for looping over specific multi-dimensional indicies in ``for`` loops.
+
+    # TODO: how do I get this to show up in the docs
+    class ndrange(stop)
+    class ndrange(start, stop[, step])
+
+    The arguments to the ``ndrange`` constructor must be iterables (lists,
+    tuple, 1D-arrays, etc.) of integers. Individual integers are also accepted
+    and are coerced to tuples of length 1.
+
+    The length of all provided arguments must be equal.
+
+    While conceptually similar to a multidimensional object, the iterator
+    returned by ndrange is a flat iterator that would be determined by the
+    array of tuples
+
+      r[i_0, i_1, ..., i_Nminus1] = (start[0] + step[0] * i_0,
+                                     start[1] + step[1] * i_1,
+                                     ...,
+                                     start[N-1] + step[N-1] * i_Nminus1)
+
+    where ``i0, i1, ..., iN_m1 >= 0``. For positive steps, at an axis ``j``,
+    the largest index is constrained such that
+    ``r[..., i_j, ...][j] < stop[j]``. For negative steps at an axis ``j``,
+    sthe smallest index at that axis is contrained by
+    ``r[..., i_j, ...][j] > stop[j]``.
```

Comment:
On first readthrough I find this is a bit confusing. What is the object `r`? After re-reading, it is an array of tuples, however that's a strange case since normally you can't have arrays of tuples, except for object arrays.

What about, instead of explaining using an "array of tuples", saying something like 
```
When iterated, ndrange produces a sequence of tuples of integers 
`(i_0, i_1, ..., i_Nminus1)` as produced by nested for-loops over ranges. For instance,
for dimension 3 this sequence is identical to that produced by:

    ((x,y,z) for x in range(start[0], stop[0], step[0])
             for y in range(start[1], stop[1], step[1]) 
             for z in range(start[2], stop[2], step[2]))
```

Reply:
Yeah, I like your wording more.

FYI, now that `__array__` has been implemented, the conceptual matrix is the ndarray object itself. technically it can be created. by calling `np.asarray(ndrange((3, 4))`. Not too sure how to document that.